HJM.Chip8/HJM.Chip8.CPU/CPUState.cs

```diff
@@ -0,0 +1,141 @@
+        /// 0x200-0xFFF - Program ROM and work RAM
+        /// </summary>
+        public byte[] Memory { get; set; } = new byte[4096];
+        public byte[] Registers { get; set; } = new byte[16];
+        public ushort IndexRegister { get; set; }
+        public ushort ProgramCounter { get; set; }
+        public byte[] Graphics { get; set; } = new byte[64 * 32];
+        public byte DelayTimer { get; set; }
+        public byte SoundTimer { get; set; }
+        public ushort[] Stack { get; set; } = new ushort[16];
+        public ushort StackPointer { get; set; }
+        public byte[] Key { get; set; } = new byte[16];
+
+        /// <summary>
+        /// Apply the given state change
+        /// </summary>
+        /// <param name="stateChange">state change to apply</param>
+        public void ApplyStateChange(CPUStateChange stateChange)
+        {
+            foreach (AddressChange<byte> registerChange in stateChange.RegisterChanges)
+            {
+                Registers[registerChange.AddressChanged] = registerChange.NewValue;
+            }
+
+            foreach (AddressChange<byte> memoryChange in stateChange.MemoryChanges)
+            {
+                Memory[memoryChange.AddressChanged] = memoryChange.NewValue;
+            }
+
+            foreach (AddressChange<byte> graphicChange in stateChange.GraphicsChanges)
+            {
+                Graphics[graphicChange.AddressChanged] = graphicChange.NewValue;
+            }
+
+            if (stateChange.ProgramCounterChange != null)
+            {
+                ProgramCounter = stateChange.ProgramCounterChange.NewValue;
+            }
+
+            if (stateChange.IndexRegisterChange != null)
+            {
+                IndexRegister = stateChange.IndexRegisterChange.NewValue;
+            }
+
+            if (stateChange.StackChange != null)
+            {
+                if (stateChange.StackChange.StackPointerChange != null)
+                {
+                    StackPointer = stateChange.StackChange.StackPointerChange.NewValue;
+                }
+
+                if (stateChange.StackChange.AddressStackChange != null)
+                {
+                    Stack[stateChange.StackChange.AddressStackChange.AddressChanged] = stateChange.StackChange.AddressStackChange.NewValue;
+                }
+            }
+
+            if (stateChange.SoundTimerChange != null)
+            {
+                SoundTimer = stateChange.SoundTimerChange.NewValue;
+            }
+
+            if (stateChange.DelayTimerChange != null)
+            {
+                DelayTimer = stateChange.DelayTimerChange.NewValue;
+            }
+        }
+
+        /// <summary>
+        /// Revert the given state change
+        /// </summary>
+        /// <param name="stateChange">State change to revert</param>
+        public void RevertStateChange(CPUStateChange stateChange)
+        {
+            foreach (AddressChange<byte> registerChange in stateChange.RegisterChanges)
```

Comment:
Consider splitting up these steps to own methods. It would make the code more readable when method name would indicate what is happening.